react/game_area.js

```diff
@@ -134,6 +141,87 @@ class GameArea extends Component {
       team1Completed: json.team1Completed,
       team2Completed: json.team2Completed
     });
+    this.checkFinishedEnvelopes();
+  }
+
+  checkFinishedEnvelopes(){
+    let finishedTemp = this.state.finishedEnvelopes;
+    let prevCompleted = this.state.envelopes.filter(item => item.prevCompleted === true)
+    
+    for (let i = 0; i < prevCompleted.length; i++) {
+        if (this.state.finishedEnvelopes.has(prevCompleted[i].envelopeId) && prevCompleted[i].seatNumber === 3) {
+          finishedTemp.delete(prevCompleted[i].envelopeId);
+          
+        }
+      this.setState({finishedEnvelopes: finishedTemp});
+    }
+  }
+
+  async finishActiveEnvelope() {
+    let seat = this.state.seats ? this.state.seats.find(i => {
+      return i.seatId === this.state.seatId;
+    }) : null;
+    let finished = this.state.finishedEnvelopes;
+    finished.add(this.state.activeEnvelope.envelopeId);
+    if (seat.isTeam1) {
+      this.advanceEnvelopeSeat([this.state.activeEnvelope.envelopeId]);
+    } else {
+      let envelope = this.state.activeEnvelope;
+      envelope.clientState = 5;
+      await this.updateActiveEnvelope(envelope);
+    }
+    this.setState({ activeEnvelope: null, finishedEnvelopes: finished });
+  }
+
+  setActiveEnvelope() {
+    let envelopes=this.state.envelopes ? this.state.envelopes.filter((i) => {
+      return i.seatNumber === this.state.mySeatNumber && i.isTeam1 === this.state.isTeam1
+    }) : [];
+    if (envelopes && this.state.activeEnvelope === null && envelopes.length > 0) {
+      // find an envelope that is not the active envelope and is not in the finished envelope set
+      let unfinishedEnvelopes = envelopes.filter((e) => {
+        return !this.state.finishedEnvelopes.has(e.envelopeId);
+      });
+      if (unfinishedEnvelopes.length > 0) {
+        unfinishedEnvelopes[0].complete = false;
+        unfinishedEnvelopes[0].stamped = false;
+        unfinishedEnvelopes[0].envelopeState = 1;
+        unfinishedEnvelopes[0].random = Array.from(Array(5), (x, i) => i + unfinishedEnvelopes[0].matchingStamp).sort(() => Math.random() - 0.5);
```

Comment:
I'd definitely try to encapsulate what's happening here into a function with a good name.  as a reviewer, I have no idea what's going on.